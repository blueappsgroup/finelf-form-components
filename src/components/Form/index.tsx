import React, { FC, ReactElement, useMemo, useState } from 'react'
import { Formik, Form } from 'formik'
import styled from 'styled-components'

import { ThemeProvider } from '../../consts/theme'
import { device } from '../../consts/sizes'
import { FormProps, FormValuesType } from '../../types'
import RedirectPage from '../RedirectPage'
import {
  FormContext,
  getFormValuesFromCache,
  resetFormValueCache,
  handleSendDataToApi,
  getFieldsValuesFromUrl,
} from '../../utils'
import { formStatuses } from '../../consts/form'

const StyledForm = styled(Form)`
  display: flex;
  flex-direction: column;
  max-width: ${(props): string => props.theme.formMaxWidth};
  justify-self: center;
  background: ${(props): string => props.theme.formBgColor};
  font-family: ${(props): string => props.theme.fontFamily};
  border-radius: 6px;
  box-shadow: ${(props): string => props.theme.formBoxShadow};
`

const FormWrapper: FC<FormProps> = ({
  children,
  onSubmit,
  customTheme,
  id,
  stepsLength,
  stepsTitles,
  hasRedirect,
  redirectUrl,
  timeToRedirect,
  redirectHeaderText,
  logoImg,
  redirectMainImg,
  redirectBgImg,
  sendDataToApi,
  apiUrl,
  transactionName,
  propertyNamesFromUrl,
  dataWithUserAgent,
}) => {
  const trasationIdValue =
    transactionName &&
    new URLSearchParams(window.location.search).get(transactionName)
  const intialValuesFromUrl = useMemo(
    () =>
      (propertyNamesFromUrl &&
        propertyNamesFromUrl.length > 0 &&
        getFieldsValuesFromUrl(propertyNamesFromUrl)) ||
      {},
    [propertyNamesFromUrl]
  )
  const [redirectUrlPath, setRedirectUrlPath] = useState(redirectUrl)
  const [initialValues, setInitialValues] = useState({
    ...intialValuesFromUrl,
    ...getFormValuesFromCache(id),
    // eslint-disable-next-line @typescript-eslint/camelcase
    trasaction_id: trasationIdValue,
  })
  const [currentStep, setCurrentStep] = useState(0)
  const [fieldsForSkip, setFieldsForSkip] = useState<string[]>([])
  const [errorFromApi, setErrorFromApi] = useState<boolean>(false)
  const shouldRedirect = !errorFromApi && hasRedirect
  const addFieldForSkip = (key: string): void =>
    setFieldsForSkip([...fieldsForSkip, key])

  const handleSubmit = async (
    values: FormValuesType,
    props: {
      resetForm: () => void
      setStatus: (status: string) => void
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ): Promise<any> => {
    if (onSubmit) {
      onSubmit(values, props)
    }

    try {
      if (sendDataToApi && apiUrl) {
        const response = await handleSendDataToApi(
          values,
          apiUrl,
          id,
          fieldsForSkip,
          dataWithUserAgent
        )
        const {
          redirectUrl: urlFromApi,
          status: statusFromApi,
        } = await response.json()

        if (statusFromApi === false) {
          setErrorFromApi(true)
          return
        }

        urlFromApi && setRedirectUrlPath(urlFromApi)
      }

      props.resetForm()
      props.setStatus(formStatuses.submited)
    } catch (e) {
      props.setStatus(formStatuses.error)
    }
  }
  const handleReset = (): void => {
    resetFormValueCache(id)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    setInitialValues({} as any)
  }

  const prevStep: Function = () => setCurrentStep(currentStep - 1)

  const nextStep: Function = () => setCurrentStep(currentStep + 1)

  return (
    <FormContext.Provider
      value={{
        id,
        apiUrl,
        stepsLength,
        currentStep,
        stepsTitleList: stepsTitles,
        nextStep,
        prevStep,
        fieldsForSkip,
        addFieldForSkip,
        errorFromApi,
      }}
    >
      <ThemeProvider customTheme={{ ...customTheme }}>
        <Formik
          enableReinitialize
          initialValues={initialValues}
          onSubmit={handleSubmit}
          onReset={handleReset}
        >
          {(props): ReactElement =>
            (shouldRedirect && props.status === formStatuses.submited && (
              <RedirectPage
                redirectUrl={redirectUrlPath}
                backgroundImage={redirectBgImg}
                logoImg={logoImg}
                headerText={redirectHeaderText}
                timeToRedirect={timeToRedirect}
                mainImg={redirectMainImg}
              />
            )) || (
              <StyledForm id={id}>
                {children}
              </StyledForm>
            )
          }
        </Formik>
      </ThemeProvider>
    </FormContext.Provider>
  )
}

export default FormWrapper
