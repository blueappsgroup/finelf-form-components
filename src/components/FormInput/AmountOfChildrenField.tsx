import React, { ReactElement } from 'react'
import { Field } from 'formik'

import BaseField from './base'
import { validateAmountOfChildren } from './validate'
import { AmountOfChildrenFieldProps } from '../../types'

const AmountOfChildrenField: (
  props: AmountOfChildrenFieldProps
) => ReactElement = ({
  id,
  label,
  placeholder = 'AmountOfChildren',
  errorMsg = 'Podana ilość jest nieprawidłowa',
  name = 'amount_of_children',
  required = false,
  showError,
}) => (
  <Field
    required={required}
    showError={showError}
    id={id}
    name={name}
    component={BaseField}
    label={label}
    placeholder={placeholder}
    type="text"
    validate={validateAmountOfChildren(errorMsg, required)}
  />
)

export default AmountOfChildrenField
