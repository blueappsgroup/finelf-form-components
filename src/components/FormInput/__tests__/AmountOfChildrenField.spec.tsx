import * as React from 'react'
import { render, fireEvent, act } from '@testing-library/react'

import Form from '../../Form'
import AmountOfChildrenField from '../AmountOfChildrenField'

describe('<AmountOfChildrenField />', () => {
  const onSubmit = jest.fn()
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const setupWrapper: (config: any) => any = ({
    formId = 'testForm',
    inputId = 'amountOfChildrenField',
    inputName = 'name',
    required = false,
    validate,
    ...rest
  }) => {
    const wrapper = render(
      <Form id={formId} onSubmit={onSubmit}>
        <AmountOfChildrenField
          id={inputId}
          name={inputName}
          validate={validate}
          required={required}
          {...rest}
        />
      </Form>
    )
    const { container } = wrapper

    const input = container.querySelector(`[name="${inputName}"]`)

    return {
      input,
      ...wrapper,
    }
  }

  it('matches snapshot', () => {
    const { input } = setupWrapper({})

    expect(input).toBeTruthy()
    expect(input).toMatchSnapshot()
  })

  it('matches snapshot with label', () => {
    const wrapper = setupWrapper({ label: 'Test' })
    const { container } = wrapper

    expect(wrapper.getByText('Test')).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it('matches snapshot with error', async () => {
    const wrapper = setupWrapper({
      label: 'Test',
      showError: true,
      required: true,
      error: 'true',
    })
    const { container, input } = wrapper

    await act(async () => {
      fireEvent.focus(input, { target: { value: '0' } })
    })

    await act(async () => {
      fireEvent.blur(input, { target: { value: '0' } })
    })

    expect(wrapper.getByText('To pole jest wymagane')).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it('changes input field', async () => {
    const { input } = setupWrapper({})

    await act(async () => {
      fireEvent.change(input, { target: { value: '0' } })
    })

    expect(input.value).toBe('0')
  })

  it('input field onBlur save value to sessionStorage', async () => {
    const { input } = setupWrapper({})

    await act(async () => {
      fireEvent.change(input, { target: { value: '0' } })
    })

    await act(async () => {
      fireEvent.blur(input, { target: { value: '0' } })
    })

    expect(
      JSON.parse(global.window.sessionStorage.getItem('form-testForm')).name
    ).toBe('0')
    expect(input.value).toBe('0')
  })

  it('input field with custom validation', async () => {
    const customValidate = () => (value: string): string => {
      if (value === 'test') {
        return 'Podana ilość jest nieprawidłowa'
      }
    }
    const wrapper = setupWrapper({
      validate: customValidate,
      showError: true,
    })
    const { input, container } = wrapper

    await act(async () => {
      fireEvent.change(input, { target: { value: 'test' } })
    })

    await act(async () => {
      fireEvent.blur(input, { target: { value: 'test' } })
    })

    expect(container).toMatchSnapshot()
    expect(wrapper.getByText('Podana ilość jest nieprawidłowa')).toBeTruthy()
  })
})
