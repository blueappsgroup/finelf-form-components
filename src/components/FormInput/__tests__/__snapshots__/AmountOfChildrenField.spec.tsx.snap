// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AmountOfChildrenField /> input field with custom validation 1`] = `
<div>
  <form
    action="#"
    class="sc-fujyUd horYMd"
    id="testForm"
  >
    <input
      class="sc-iCoHVE gEjxKq"
      form="[object Object]"
    />
    <div
      class="sc-pNWxx iTQnTB"
    >
      <div
        class="sc-dIvqjp iaFmok"
      >
        <input
          class="sc-crzoUp jQLlWv"
          id="amountOfChildrenField"
          name="name"
          placeholder="Liczba osób na utrzymaniu"
          type="text"
          value="test"
        />
      </div>
      <span
        class="sc-iqAbSa dmIjtF"
      >
        Podana ilość jest nieprawidłowa
      </span>
    </div>
  </form>
</div>
`;

exports[`<AmountOfChildrenField /> matches snapshot 1`] = `
<input
  class="sc-crzoUp kUQhja"
  id="amountOfChildrenField"
  name="name"
  placeholder="Liczba osób na utrzymaniu"
  type="text"
  value=""
/>
`;

exports[`<AmountOfChildrenField /> matches snapshot with error 1`] = `
<div>
  <form
    action="#"
    class="sc-fujyUd horYMd"
    id="testForm"
  >
    <input
      class="sc-iCoHVE gEjxKq"
      form="[object Object]"
    />
    <div
      class="sc-pNWxx iTQnTB"
    >
      <label
        for="name"
      >
        Test*
      </label>
      <div
        class="sc-dIvqjp iaFmok"
      >
        <input
          class="sc-crzoUp jQLlWv"
          id="amountOfChildrenField"
          label="Test"
          name="name"
          placeholder="Liczba osób na utrzymaniu*"
          required=""
          type="text"
          value=""
        />
      </div>
      <span
        class="sc-iqAbSa dmIjtF"
      >
        To pole jest wymagane
      </span>
    </div>
  </form>
</div>
`;

exports[`<AmountOfChildrenField /> matches snapshot with label 1`] = `
<div>
  <form
    action="#"
    class="sc-fujyUd horYMd"
    id="testForm"
  >
    <input
      class="sc-iCoHVE gEjxKq"
      form="[object Object]"
    />
    <div
      class="sc-pNWxx iTQnTB"
    >
      <label
        for="name"
      >
        Test
      </label>
      <div
        class="sc-dIvqjp iaFmok"
      >
        <input
          class="sc-crzoUp kUQhja"
          id="amountOfChildrenField"
          label="Test"
          name="name"
          placeholder="Liczba osób na utrzymaniu"
          type="text"
          value=""
        />
      </div>
    </div>
  </form>
</div>
`;
